plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java

```diff
@@ -79,53 +73,66 @@ public static void addMavenLibraries(IProject project, List<Library> libraries,
         Messages.getString("adding.app.engine.libraries"), //$NON-NLS-1$
-        libraries.size() + 1); // + 1 because we pass the submonitor along below
-
-    List<IClasspathEntry> rawClasspath = Lists.newArrayList(javaProject.getRawClasspath());
+        2);
+    
+    Library masterLibrary = collectLibraryFiles(javaProject, libraries);
+    
+    List<IClasspathEntry> newEntries = computeEntries(javaProject, masterLibrary);
+    subMonitor.worked(1);
+    
+    ClasspathUtil.addClasspathEntries(javaProject.getProject(), newEntries, subMonitor);
+    runContainerResolverJob(javaProject);
+    
+    return newEntries.toArray(new IClasspathEntry[0]);
+  }
 
-    List<IClasspathEntry> newEntries = new ArrayList<>();
+  /**
+   * Adds all jars and dependencies from <code>libraries</code> to the master library.
+   * 
+   * @return the master library
+   */
+  public static Library collectLibraryFiles(IJavaProject javaProject, List<Library> libraries)
+      throws CoreException {
+    SortedSet<LibraryFile> masterFiles = new TreeSet<>();
     for (Library library : libraries) {
-      IClasspathEntry libraryContainer = makeClasspathEntry(library);
-      if (!rawClasspath.contains(libraryContainer)) {
-        newEntries.add(libraryContainer);
+      if (!library.isResolved()) {
+        library.resolveDependencies();
       }
-      subMonitor.worked(1);
+      masterFiles.addAll(library.getLibraryFiles());
     }
 
-    if (addToClasspath) {
-      ClasspathUtil.addClasspathEntries(javaProject.getProject(), newEntries, subMonitor);
-      runContainerResolverJob(javaProject);
-    }
-
-    return newEntries.toArray(new IClasspathEntry[0]);
+    Library masterLibrary = CloudLibraries.getMasterLibrary(javaProject);
+    masterFiles.addAll(masterLibrary.getLibraryFiles());
+    masterLibrary.setLibraryFiles(new ArrayList<LibraryFile>(masterFiles));
+    return masterLibrary;
   }
 
+  private static List<IClasspathEntry> computeEntries(IJavaProject javaProject, Library library)
+      throws CoreException {
+    List<IClasspathEntry> rawClasspath = Lists.newArrayList(javaProject.getRawClasspath());
+    List<IClasspathEntry> newEntries = new ArrayList<>();
+    IClasspathEntry libraryContainer = makeClasspathEntry(library);
+    if (!rawClasspath.contains(libraryContainer)) {
+      newEntries.add(libraryContainer);
+    }
+    return newEntries;
+  }
+  
   /**
-   * @return the entries to be added to the classpath. Does not add them to the classpath.
+   * Returns the entries to be added to the classpath. Does not add them to the classpath.
    */
-  public static IClasspathEntry[] listAdditionalLibraries(
-      IJavaProject javaProject, List<Library> libraries, IProgressMonitor monitor)
-          throws JavaModelException, CoreException {
-    return prepareLibraries(javaProject, libraries, monitor, false);
+  public static IClasspathEntry[] listNativeLibrary(IJavaProject javaProject, Library library)
```

Comment:
Similarly, this really means `listNativeMasterLibrary()`, returning either a singleton array or an empty list.

Reply:
This one, I think, operates on the argument, not the master library.

Reply:
That's true. And at the current state, the argument `library` is always `masterLibrary`, so I'm also implying we could remove the argument going forward. However, let's not overthink it.